subsys/modem/modem_chat.c

```diff
@@ -774,6 +774,7 @@ int modem_chat_run_script_async(struct modem_chat *chat, const struct modem_chat
 		return -EBUSY;
 	}
 
+	modem_chat_parse_reset(chat);
```

Comment:
This is not ideal since the parser doesn't only parse scripts, it also parses URCs, which this call may corrupt, possibly catastrophically since internal members of modem_chat must not be modified by other threads than the sys workqueue without additional thread synchronization :)

The best approach to identifying and solving the possible issue causing issues in your usecase is to add a test to the unit test suite which induces the errors you are seeing, that really helps identify issues since they are run in native_sim, giving us full thread awareness while debugging :)

See https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/subsys/modem/modem_chat/src/main.c

Reply:
So, does this mean it is possible that a chat script can be running even if MODEM_CHAT_SCRIPT_STATE_RUNNING_BIT is not set? That seems like a much bigger issue

Reply:
Nope :) The parser is always running, reading all received lines, and trying to match them against all currently active matches. These could be URCs, set in the init function, and matches contained within the currently running script, if any.

See 
https://github.com/zephyrproject-rtos/zephyr/blob/6a42ede43b53ceffddeaa814b4bd99284c25240d/drivers/gnss/gnss_quectel_lcx6g.c#L123-L129
and
https://github.com/zephyrproject-rtos/zephyr/blob/6a42ede43b53ceffddeaa814b4bd99284c25240d/drivers/gnss/gnss_quectel_lcx6g.c#L728-L729

These are always active, and the parser is looking for any matching lines, even when a script is not running :)

Scripts can temporarily add additional matches to the modem_chat instance, which are only relevant while the script is running, and are removed once the script either stops are moves to the next script command.